### PR TITLE
Keyboard - Ginny from gBoards

### PR DESCRIPTION
**Pull request by germ**
A fully functional ASETNIOP keyboard with a onboard chording engine that uses generated keymaps.

This PR adds support for this keyboard 

- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)

### CODE REVIEW

**Comment by germ** [comment]
Ack, accidently bundled in some stuff that's not supposed to be here. Give me a few

**Comment by germ** [comment]
Should be good now. Don't look at how the sausage is made.

File: keyboards/ginny/config.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+// Copy and worked on with love from the EZ team
+
+#pragma once
+#include "config_common.h"
+
+// Space saving jazz
+#define NO_ACTION_LAYER
+#define NO_ACTION_TAPPING
+#define NO_ACTION_ONESHOT
```
Comment (drashna) [inline keyboards/ginny/config.h]: I'm not a fan of adding this, but .... yeah super custom.

File: keyboards/ginny/matrix.c [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+    matrix_scan_quantum();
+
+#ifdef DEBUG_MATRIX
+    for (uint8_t c = 0; c < MATRIX_COLS; c++)
+		for (uint8_t r = 0; r < MATRIX_ROWS; r++)
+		  if (matrix_is_on(r, c)) xprintf("r:%d c:%d \n", r, c);
+#endif
+
+    return 1;
```
Comment (drashna) [inline keyboards/ginny/matrix.c]:
```suggestion
    return changed;
```

File: keyboards/ginny/readme.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+[More info on qmk.fm](https://www.gboards.ca/product/ginni/)
+
+Keyboard Maintainer: [Jeremy Bernhardt](https://github.com/germ)  
+Hardware Supported: Gergo (Kit, Partial, Ready)   
+Hardware Availability: [gboards.ca](http://gboards.ca)  
+
+## Firmware building
+After cloning the QMK repo and installing dfu-programmer build and flash with. Be sure to reset your keyboard!
+
+    make ginny:default:dfu
```
Comment (drashna) [inline keyboards/ginny/readme.md]:
```suggestion
    make ginny:default:flash
```

File: keyboards/ginny/readme.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Ginny!
+
+![Ginny](https://assets.bigcartel.com/product_images/247416878/IMG_20191025_1424438-01.jpeg)
+
+Quite possibly the smallest most feature filled board you can fit in your pocket. With multi-language
+onboard ASETNIOP and all sorts of fun shenanigans
+
+[More info on qmk.fm](https://www.gboards.ca/product/ginni/)
+
+Keyboard Maintainer: [Jeremy Bernhardt](https://github.com/germ)
```
Comment (drashna) [inline keyboards/ginny/readme.md]:
```suggestion
* Keyboard Maintainer: [Jeremy Bernhardt](https://github.com/germ)  
```

File: keyboards/ginny/readme.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+![Ginny](https://assets.bigcartel.com/product_images/247416878/IMG_20191025_1424438-01.jpeg)
+
+Quite possibly the smallest most feature filled board you can fit in your pocket. With multi-language
+onboard ASETNIOP and all sorts of fun shenanigans
+
+[More info on qmk.fm](https://www.gboards.ca/product/ginni/)
+
+Keyboard Maintainer: [Jeremy Bernhardt](https://github.com/germ)  
+Hardware Supported: Gergo (Kit, Partial, Ready)
```
Comment (drashna) [inline keyboards/ginny/readme.md]:
```suggestion
* Hardware Supported: Gergo (Kit, Partial, Ready)   
```

File: keyboards/ginny/readme.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+![Ginny](https://assets.bigcartel.com/product_images/247416878/IMG_20191025_1424438-01.jpeg)
+
+Quite possibly the smallest most feature filled board you can fit in your pocket. With multi-language
+onboard ASETNIOP and all sorts of fun shenanigans
+
+[More info on qmk.fm](https://www.gboards.ca/product/ginni/)
+
+Keyboard Maintainer: [Jeremy Bernhardt](https://github.com/germ)  
+Hardware Supported: Gergo (Kit, Partial, Ready)   
+Hardware Availability: [gboards.ca](http://gboards.ca)
```
Comment (drashna) [inline keyboards/ginny/readme.md]:
```suggestion
* Hardware Availability: [gboards.ca](http://gboards.ca)  
```

File: keyboards/ginny/rules.mk [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+# Lets try and save some space eh?
+MOUSEKEY_ENABLE   = no
+EXTRAKEY_ENABLE   = no
+CONSOLE_ENABLE		= no
+COMMAND_ENABLE		= no
+
+SRC							+= matrix.c sten.c
+QUANTUM_LIB_SRC += i2c_master.c
+OPT_DEFS				+= -DONLYQWERTY -DDEBUG_MATRIX
+EXTRAFLAGS			+= -flto
```
Comment (drashna) [inline keyboards/ginny/rules.mk]:
```suggestion
LTO_ENABLE = yes
```

File: keyboards/ginny/config.h (hunk 1 above)
Comment (drashna) [reply keyboards/ginny/config.h]: Also, could add `#define LAYER_STATE_8BIT` to use an 8 bit bitmask, instead.

**Comment by germ** [comment]
Minor fix in how layers/chord conflicts are handled.

**Comment by germ** [comment]
We good to merge into master?

File: keyboards/ginny/ginny.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
+#include QMK_KEYBOARD_H
+
+bool i2c_initialized = 0;
+i2c_status_t mcp23018_status = 0x20;
+
+void matrix_init_kb(void) {
+    // (tied to Vcc for hardware convenience)
```
Comment (Duckle29) [inline keyboards/ginny/ginny.c]: Just a note that we do have pin functions in qmk. It won't be as compact, but it'd be more readable. for example:

```C
setPinInputHigh(B0);
setPinInputHigh(B4);
setPinInputHigh(B5);
setPinInputHigh(B6);
setPinInputHigh(B7);
```
For the portB pins.

Full list here: https://beta.docs.qmk.fm/for-makers-and-modders/internals_gpio_control#functions

**Comment by germ** [comment]
Moved changes into  #8262 , closing.

**Pull request closed without merging**